design/pipeline-caching.md

```diff
@@ -0,0 +1,154 @@
+
+Selecting the cache strategy would simply be a matter of specifying a strategy on the task inputs (for both the restore and save tasks, restore shown below):
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    strategy: dedup
+    key: |
+      package.json
+      yarn.lock
+    path: node_modules
+```
+
+### YAML syntax
+
+The example task usage above uses the explicit task references. We also want to provide a streamlined syntax for caching, applied to the example above.
+
+```yaml
+steps:
+- restoreCache: yarn.lock
+- script: yarn
+- saveCache:
+  key: yarn.lock
+  path: node_modules
+- script: yarn build
+```
+
+This more compact syntax just reduces some of the noise around using the caching task. Additionally we are planning to integrate caching into the the ```use:``` syntax. This will look like the following:
+
+```yaml
+steps:
+- use: node
+  cache: true
+- script: |
+    yarn
+    yarn build
+    yarn test
+```
+
+The idea behind the ```use:``` syntax is that given what we are using (in this case ```node```) we make some automatic decisions about what and how to cache. This logic will use various heuristics to make the decision about what to cache, for example, it will look for the presence of a ```yarn.lock``` file or a ```package-lock.json``` to determine where the ```node_modules``` path may be located.
+
+Other ```use:``` statements will have similar heuristics to pick the best defaults.
+
+Note that the ```use:``` syntax will inject the cache save step at the end of the build process which will not always be desirable. In those cases developers can fall back to specifying the ```restoreCache:``` and ```saveCache:``` YAML statements.
+
+### Cache Scoping
+
+Getting scoping right is important for maximising cache hits and but also avoiding the cache becoming an attack vector to insert malicous code into official/master builds. A cache scope will be defined by the pipeline ID, the Git repos and refs that are being built and any other special considerations - such as whether the build is for a particular pull request. For regular builds on a branch (non PR scenario) a single scope will be used and the job will have read/write access to this scope.
+
+For PR builds it is a little bit more complex because we want the PR build to be able to use the scope for the source or target branch of it is available. So i nthe PR scenario the job will have access to three scopes, one that represents the target branch, one that represents the source branch (these will both be R/O) and a third which represents the PR specifically. The PR scope will be tried first, then the source branch and then the target branch. In general topic-branch workflows the first build on a PR branch will be a cache miss on PR scope, the source branch will also likely be a miss but the third scope should be a hit (the target branch). The reason for these precedence rules is to allow for the most specific cache to be used.
```

Comment:
Given that the source and target branch scopes are considered read-only from the PR lifetime point of view (i.e. creating, updating, abandoning a PR), what sort of process would take care of creating the source and target scope caches? Would it be something like a background job?

Reply:
@juliobbv the target scope would be populated by a subsequent build on that branch. For example you may have a CI build that builds master, and associates a cache entry based on the contents of the package-lock.json file (in an NPM example).

The PR build would be able to use this cache *if* the package-lock.json file hasn't changed. The assumption here is that the target branch is a well protected branch with code reviews/branch policies etc.

The source branch scenario is probably going to be less common, but we do see folks setting up CI-style builds for longer lived feature branches, so in this scenario the PR build would be able to use the cache produced from those CI builds.

With designer-based Pipelines its more common that you have a seperate definition for CI builds and PR builds, but with YAML builds its more likely to share the same pipeline definition across all the build scenarios (and use templates to pull in different functionality for each scenario).

Reply:
Makes sense @mitchdenny, I got it!